src/path/parse.ts

```diff
@@ -529,21 +529,201 @@ export class Parser {
     return left;
   }
 
-  protected decodeString(token: Token, isName: boolean = false): string {
-    try {
-      return JSON.parse(
-        token.kind === TokenKind.SINGLE_QUOTE_STRING
-          ? `"${token.value.replaceAll('"', '\\"').replaceAll("\\'", "'")}"`
-          : `"${token.value}"`,
-      );
-    } catch {
-      throw new JSONPathSyntaxError(
-        `invalid ${isName ? "name selector" : "string literal"} '${
-          token.value
-        }'`,
-        token,
-      );
+  protected decodeString(token: Token): string {
+    return this.unescapeString(
+      token.kind === TokenKind.SINGLE_QUOTE_STRING
+        ? token.value.replaceAll('"', '\\"').replaceAll("\\'", "'")
+        : token.value,
+      token,
+    );
+  }
+
+  // eslint-disable-next-line sonarjs/cognitive-complexity
+  protected unescapeString(value: string, token: Token): string {
+    const rv: string[] = [];
+    const length = value.length;
+    let index = 0;
+    let digits: string;
+    let codepoint: number;
+    let lowSurrogate: number;
+    let unescaped: string;
+    let codepointAt: number | undefined;
+
+    while (index < length) {
+      const ch = value[index];
+      if (ch === "\\") {
+        // Handle escape sequences
+        index += 1;
+
+        switch (value[index]) {
+          case '"':
+            rv.push('"');
+            break;
+          case "\\":
+            rv.push("\\");
+            break;
+          case "/":
+            rv.push("/");
+            break;
+          case "b":
+            rv.push("\x08");
+            break;
+          case "f":
+            rv.push("\x0C");
+            break;
+          case "n":
+            rv.push("\n");
+            break;
+          case "r":
+            rv.push("\r");
+            break;
+          case "t":
+            rv.push("\t");
+            break;
+          case "u":
+            if (index + 4 >= length) {
+              throw new JSONPathSyntaxError(
+                `invalid escape sequence at offset ${index}`,
+                token,
+              );
+            }
+
+            index += 1;
+
+            digits = value.slice(index, index + 4);
+            codepoint = this.parseInt16(digits, token);
+
+            if (isNaN(codepoint)) {
+              throw new JSONPathSyntaxError(
+                `invalid escape sequence at offset ${index}`,
+                token,
+              );
+            }
+
+            if (
+              index + 5 < length &&
+              value[index + 4] === "\\" &&
+              value[index + 5] === "u"
+            ) {
```

Comment:
Shouldn't this first check if `isHighSurrogate(codepoint)` before doing any further validation? Otherwise, wouldn't something like `\u1234\u1234` erroneously lead to an error here?

Should the logic here rather look like this (pseudo-code)?
```typescript
if (isHighSurrogate(codepoint)) {
  lowSurrogate = ...;
  if (!isLowSurrogate(codepoint)) {
    throw error;
  }
  
  codepoint = ...; // combine high surrogate and low surrogate
} else if (isLowSurrogate(codepoint)) {
  throw error; // unpaired low surrogate
}

... // push codepoint to string
```